SWC-7242

nickgros opened the pull request:
- Add Vite integration to bundle and emit assets
- Update HtmlInjectionFilter to inject a CDN endpoint and Vite asset tags
- Update Portal.html to use injected CDN endpoint and injected Vite asset tags
- Add script that bundles React, ReactDOM, ReactQuery, SRC, MUI components and adds them to the window object (replacing UMD scripts in Portal.html)
- Migrate from yarn to pnpm
- Update dependencies to match those that are actually used (required by pnpm) and remove some that are unused.

Depends on a synapse-react-client release containing https://github.com/Sage-Bionetworks/synapse-web-monorepo/pull/1565

nickgros committed:
SWC-7242

- Add Vite integration to bundle and emit assets
- Update HtmlInjectionFilter to inject a CDN endpoint and Vite asset tags
- Update Portal.html to use injected CDN endpoint and injected Vite asset tags
- Add script that bundles React, ReactDOM, ReactQuery, SRC, MUI components and adds them to the window object (replacing UMD scripts in Portal.html)
- Migrate from yarn to pnpm
- Update dependencies to match those that are actually used (required by pnpm) and remove some that are unused.

nickgros commented on js/main.js:
```diff
@@ -0,0 +1,26 @@
+/**
```
Vite will bundle this file, and the bundled asset will be emitted in the Portal HTML.

This will replace the various UMD bundles that many packages are no longer providing.

nickgros commented on package.json:
```diff
@@ -54,27 +64,35 @@
+    "build:js": "pnpm build:js:vite",
+    "build:js:vite": "vite build",
+    "build:js:esbuild": "esbuild js/**.ts --bundle --platform=browser --format=iife --outdir=./src/main/webapp/generated",
+    "build:js:rolldown": "rolldown js/main.js --platform browser --format esm --file ./src/main/webapp/generated/swc-modules.js",
```
```suggestion
```

nickgros commented on package.json:
```diff
@@ -1,6 +1,13 @@
 {
```
While it looks like we added a bunch of new dependencies, most of these are _not_ new. With yarn v1, we could directly depend on artifacts from transitive dependencies. pnpm does not let you do this, so we have to add add those previously-transitive dependencies that we currently directly depend on.

nickgros commented on package.json:
```diff
@@ -15,10 +22,11 @@
     "markdown-it-inline-comments": "^1.0.1",
     "markdown-it-math": "^4.1.1",
     "markdown-it-strikethrough-alt": "^1.0.0",
-    "markdown-it-sub": "^2.0.0",
-    "markdown-it-sup": "^2.0.0",
```
These were not used. We were already using the `-alt` packages, they just were not explicitly listed as dependencies.

nickgros commented on package.json:
```diff
@@ -54,27 +64,35 @@
     "prettier-plugin-java": "2.5.0",
     "typescript": "5.1.6",
     "uuid": "^9.0.0",
+    "vite": "^6.0.11",
+    "vite-plugin-node-polyfills": "0.17.0",
```
Add Vite

nickgros committed:
Remove unused build scripts added in iteration

nickgros commented on package.json:
```diff
@@ -54,27 +64,31 @@
     "dev:watcher": "wait-on tcp:127.0.0.1:9876 && mvn fizzed-watcher:run",
-    "dev:tomcat": "wait-on tcp:127.0.0.1:9876 && docker pull tomcat:9.0; docker run --name swc-dev --rm -p 8888:8080 -v \"/$(pwd)/target/portal-develop-SNAPSHOT/:/usr/local/tomcat/webapps/ROOT/\" -v \"/$HOME/.m2/settings.xml\":/root/.m2/settings.xml tomcat:9.0",
+    "dev:tomcat": "wait-on tcp:127.0.0.1:9876 && docker pull tomcat:9.0; docker run --name swc-dev --rm -p 8888:8080 -v \"/$(pwd)/target/portal-develop-SNAPSHOT/:/usr/local/tomcat/webapps/ROOT/\" -v \"/$HOME/.m2/settings.xml\":/root/.m2/settings.xml -e CATALINA_OPTS=\"-Dorg.sagebionetworks.web.client.dev.mode=true\" tomcat:9.0",
+    "dev:vite": "vite dev",
```
Add a script that runs the Vite dev server (live reloading of assets bundled by Vite)

nickgros commented on package.json:
```diff
@@ -54,27 +64,31 @@
     "dev:codeserver": "mvn clean gwt:run-codeserver",
     "dev:watcher": "wait-on tcp:127.0.0.1:9876 && mvn fizzed-watcher:run",
-    "dev:tomcat": "wait-on tcp:127.0.0.1:9876 && docker pull tomcat:9.0; docker run --name swc-dev --rm -p 8888:8080 -v \"/$(pwd)/target/portal-develop-SNAPSHOT/:/usr/local/tomcat/webapps/ROOT/\" -v \"/$HOME/.m2/settings.xml\":/root/.m2/settings.xml tomcat:9.0",
+    "dev:tomcat": "wait-on tcp:127.0.0.1:9876 && docker pull tomcat:9.0; docker run --name swc-dev --rm -p 8888:8080 -v \"/$(pwd)/target/portal-develop-SNAPSHOT/:/usr/local/tomcat/webapps/ROOT/\" -v \"/$HOME/.m2/settings.xml\":/root/.m2/settings.xml -e CATALINA_OPTS=\"-Dorg.sagebionetworks.web.client.dev.mode=true\" tomcat:9.0",
```
Add environment variable `org.sagebionetworks.web.client.dev.mode=true`, which tells the servlet that we are running in dev mode and to use the Vite dev server.

nickgros committed:
Remove copy scripts for synapse-react-client and MUI

nickgros commented on src/main/java/org/sagebionetworks/web/server/PortalServletModule.java:
```diff
@@ -71,8 +81,52 @@ public class PortalServletModule extends ServletModule {
 
-  @Override
-  protected void configureServlets() {
+  // ViteManifestProviderImpl requires ServletContext, which cannot be injected, so we create the singleton here.
+  private ViteManifestProvider viteManifestProviderSingleton;
```
Is there a better way to do this?

nickgros commented on src/main/java/org/sagebionetworks/web/server/PortalServletModule.java:
```diff
@@ -71,8 +81,52 @@ public class PortalServletModule extends ServletModule {
+  }
+
+  private void bindFilters() {
+    filter("/*").through(SSLFilter.class);
+    bind(SSLFilter.class).in(Singleton.class);
+
+    filter("*").through(HtmlInjectionFilter.class);
+    bind(HtmlInjectionFilter.class).in(Singleton.class);
+
+    filter("/*").through(GWTCacheControlFilter.class);
+    bind(GWTCacheControlFilter.class).in(Singleton.class);
+
+    filter("/js/*").through(GWTAllCacheFilter.class);
+    filter("/images/*").through(GWTAllCacheFilter.class);
+    filter("/research/*").through(GWTAllCacheFilter.class);
+    bind(GWTAllCacheFilter.class).in(Singleton.class);
+
+    filter("/*").through(JavaScriptContentTypeFilter.class);
+    bind(JavaScriptContentTypeFilter.class).in(Singleton.class);
+
+    filter("/*").through(HSTSFilter.class);
+    bind(HSTSFilter.class).in(Singleton.class);
+
+    filter("/*").through(CORSFilter.class);
+    bind(CORSFilter.class).in(Singleton.class);
+
+    filter("/*").through(XFrameOptionsFilter.class);
+    bind(XFrameOptionsFilter.class).in(Singleton.class);
```
Removed filters from web.xml and added them here, preserving the filter order. This allows us to inject dependencies into the filter objects.

nickgros commented on src/main/java/org/sagebionetworks/web/server/StackEndpoints.java:
```diff
@@ -98,6 +105,15 @@ private static String getEndpointPrefix(String host) {
         // None of these hosts match and no configuration was loaded from settings, fall back to prod
         endpointPrefix = "https://repo-prod.prod.sagebase.org";
       }
+
+      logger.warning(
+        "No configuration found for " +
+        REPO_ENDPOINT_KEY +
+        ", so servlet request will defer to the host header. Host: " +
+        host +
+        ", endpointPrefix: " +
+        endpointPrefix
+      );
```
```suggestion
```

nickgros committed:
Remove unintentional change

nickgros commented on src/main/java/org/sagebionetworks/web/server/servlet/filter/HtmlInjectionFilter.java:
```diff
@@ -149,6 +177,42 @@ public static boolean isGWTPlace(String targetPath) {
     //        return false;
   }
 
+  private void addCdnEndpoint(
+    Map<String, String> dataModel,
+    HttpServletRequest request
+  ) {
+    String origin = request.getHeader("origin");
+    if (origin != null && CDN_ORIGINS_REGEX.matcher(origin).matches()) {
+      dataModel.put(CDN_ENDPOINT_KEY, "//cdn-" + request.getHeader("origin"));
+    } else {
+      dataModel.put(CDN_ENDPOINT_KEY, "");
+    }
+  }
```
Moved logic to determine the CDN endpoint to this filter, instead of JavaScript in the HTML file itself.  Logic should be the same.

nickgros commented on src/main/java/org/sagebionetworks/web/server/servlet/filter/HtmlInjectionFilter.java:
```diff
@@ -149,6 +177,42 @@ public static boolean isGWTPlace(String targetPath) {
+    }
+  }
+
+  /**
+   * Adds Vite imports to the data model based on the request origin.
+   * This must be called AFTER the CDN_ENDPOINT_KEY is added to the data model
+   * @param dataModel
+   */
+  private void addViteImports(Map<String, String> dataModel) {
+    boolean isDev = "true".equals(System.getProperty(IS_DEV_MODE));
+    if (isDev) {
+      dataModel.put(
+        VITE_IMPORTS_INJECTION_KEY,
+        ViteHTMLProvider.getViteDevelopmentHTML(VITE_IMPORTED_FILES)
+      );
+    } else {
+      dataModel.put(
+        VITE_IMPORTS_INJECTION_KEY,
+        ViteHTMLProvider.getViteProductionHTML(
+          VITE_IMPORTED_FILES,
+          viteManifestProvider.getManifest(),
+          dataModel.get(CDN_ENDPOINT_KEY) + "/generated/vite/"
+        )
+      );
+    }
+  }
```
Injects scripts for the assets bundled by Vite

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -10,7 +10,7 @@
     <meta
       http-equiv="Content-Security-Policy"
-      content="default-src 'self' data: blob: 'unsafe-inline' 'unsafe-eval' *.synapse.org *.sagebase.org *.googleapis.com *.gstatic.com *.googletagmanager.com cdn.statuspage.io s3.amazonaws.com www.google-analytics.com kh896k90gyvg.statuspage.io *.youtube.com *.rstudio.com *.google.com *.shinyapps.io *.smartsheet.com sagebionetworks.jira.com 127.0.0.1:9876 esm.sh; frame-ancestors 'self' https://* ...
+      content="default-src 'self' data: blob: 'unsafe-inline' 'unsafe-eval' *.synapse.org *.sagebase.org *.googleapis.com *.gstatic.com *.googletagmanager.com cdn.statuspage.io s3.amazonaws.com www.google-analytics.com kh896k90gyvg.statuspage.io *.youtube.com *.rstudio.com *.google.com *.shinyapps.io *.smartsheet.com sagebionetworks.jira.com 127.0.0.1:9876 esm.sh localhost:5173; frame-ancestors ' ...
```
Vite dev server runs on `localhost:5173`, need to add to CSP to allow downloading scripts, connecting over HTTP and WebSocket protocols

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -93,126 +93,110 @@
     </script>
-    <script>
-      // determine cdn endpoint (staging or prod, do not use a cdn for local or dev builds)
-      var cdnEndpoint = '//cdn-' + location.hostname + '/'
-      var processEnvironment = 'production'
-      var synapseOrgRegex = /(www|staging|tst)\.synapse\.org$/
-
-      if (!synapseOrgRegex.test(location.hostname)) {
-        cdnEndpoint = '/'
-        processEnvironment = 'development'
-      }
```
Logic for CDN moved to the HtmlInjectionFilter. Vite should handle the `processEnvironment` logic in the bundling phase.

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -93,126 +93,110 @@
-      )
-      loadCss('https://fonts.googleapis.com/icon?family=Material+Icons')
-
-      let reactPath, reactDomPath
-
-      if (processEnvironment === 'production') {
-        reactPath = cdnEndpoint + 'generated/react.production.min.js'
-        reactDomPath = cdnEndpoint + 'generated/react-dom.production.min.js'
-      } else {
-        reactPath = cdnEndpoint + 'generated/react.development.js'
-        reactDomPath = cdnEndpoint + 'generated/react-dom.development.js'
-      }
```
React/ReactDOM scripts removed, now handled by Vite

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -93,126 +93,110 @@
-      loadJs(cdnEndpoint + 'generated/plotly.min.js')
-      loadJs(cdnEndpoint + 'generated/create-plotly-component.min.js')
-      loadJs(cdnEndpoint + 'generated/universalCookie.min.js')
-      loadJs(cdnEndpoint + 'generated/synapse-react-client.production.min.js') // We don't load SRC's scss because we include it in our own compiled scss
```
synapse-react-client import removed, now handled by Vite

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -93,126 +93,110 @@
-      loadJs(reactPath)
-      loadJs(reactDomPath)
-      loadJs(cdnEndpoint + 'generated/material-ui.production.min.js')
-      loadJs(cdnEndpoint + 'generated/react-transition-group.min.js')
```
react-transition-group bundle removed because it is unused

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -93,126 +93,110 @@
-      loadJs(cdnEndpoint + 'generated/material-ui.production.min.js')
-      loadJs(cdnEndpoint + 'generated/react-transition-group.min.js')
-      loadJs(cdnEndpoint + 'generated/prop-types.min.js')
-      loadJs(cdnEndpoint + 'generated/react-measure.js')
```
react-measure bundle removed because it is unused

nickgros commented on src/main/webapp/Portal.html:
```diff
@@ -93,126 +93,110 @@
-      loadJs(cdnEndpoint + 'generated/pica.min.js')
-      loadJs(reactPath)
-      loadJs(reactDomPath)
-      loadJs(cdnEndpoint + 'generated/material-ui.production.min.js')
```
Removed, now handled by Vite

nickgros committed:
Refactor, add tests for filter

nickgros commented on src/main/webapp/WEB-INF/web.xml:
```diff
@@ -6,94 +6,10 @@
   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd"
   version="2.5"
 >
-  <filter>
```
Filter instantiation moved to PortalServletModule so dependencies can be injected into these.

nickgros committed:
Bump actions NODE_VERSION to v20

nickgros committed:
Try Node 22, which should bundle corepack v0.31, which has the fixed keys

nickgros committed:
Remove esm.sh from CSP since it is no longer used

nickgros commented on .github/workflows/build-test-e2e.yml:
```diff
@@ -5,7 +5,7 @@ env:
   BUILD_DIR: target
   BUILD_NAME: portal-develop-SNAPSHOT.war
-  NODE_VERSION: 18.16.0
+  NODE_VERSION: 22
```
Update Node to 22 (latest LTS) since it includes the latest version of [corepack](https://nodejs.org/api/corepack.html), which we will use to install pnpm.

hallieswan reviewed (approved):
Looks great to me! It'd be great to run e2e tests locally if possible to confirm that core functionality is unaffected.

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-7242

nickgros committed:
SWC-7242

- Fix e2e tests
- Fix ReactDOM/ReactDOMClient distinction -- ReactDOMClient only contains `createRoot`
- Prefer `window` over `self`
- Move Vite compilation to `compile` phase -- it does not get copied over if we wait until the `package` phase, even though we don't need it until then.

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-7242

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-7242

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/SynapseWebClient into SWC-7242

nickgros committed:
SWC-7242

- Get Vite manifest through classloader so it will work with Jetty & Tomcat, simplifies dependency injection & setup
- Ensure Vite is built & copied before WAR is exploded
- Remove outdated context from comment
